### PR TITLE
feat(router): register NgModuleFactory objects.

### PR DESCRIPTION
**Pull request by mprobst**
When lazily loading code, users need to be able to get hold of the
NgModuleFactory. For SystemJS environments, the SystemJS registry serves
this purpose. However other environments, such as modules compiled with
Closure compiler, do not expose exports object or a path based registry.

For these environments, `@NgModule` objects can include an identifier, and
the loading code can then pass `loadModule(id).then(() =>
getNgModule(id))` to the router.

### CODE REVIEW

**Comment by mprobst** [comment]
`loadToken` is a provisional name, please suggest a better one :-)

File: modules/@angular/compiler/core_private.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
+/**
```
Comment (mhevery) [inline modules/@angular/compiler/core_private.ts]: Why is this file here? I don't think it is part of the PR

Comment (mprobst) [inline modules/@angular/compiler/core_private.ts]: Yeah, botched merge.

**Comment by mprobst** [comment]
PTAL, now with `loadToken` renamed to `id`.

File: modules/@angular/compiler/src/identifiers.ts [hunk 2]
```diff
@@ -107,6 +107,11 @@ export class Identifiers {
     runtime: NgModuleInjector,
     moduleUrl: assetUrl('core', 'linker/ng_module_factory')
   };
+  static RegisterLoadedModuleFn: IdentifierSpec = {
+    name: 'registerLoadedModule',
```
Comment (mhevery) [inline modules/@angular/compiler/src/identifiers.ts]: perhaps `registerModuleFactory` would be a better name?

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts [hunk 3]
```diff
@@ -15,3 +15,32 @@ import {NgModuleFactory} from './ng_module_factory';
...
   abstract load(path: string): Promise<NgModuleFactory<any>>;
 }
+
+let deferreds: {[id: string]: NgModuleFactory<any>} = {};
+
+/**
+ * Registers a loaded module. Should only be called from generated NgModuleFactory code.
+ * @experimental
+ */
+export function registerLoadedModule(id: string, factory: NgModuleFactory<any>) {
```
Comment (mhevery) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: consider renaming to `registerModuleFactory`?

**Comment by mhevery** [comment]
One more rename, otherwise LGTM.

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: +1

**Comment by tbosch** [comment]
Also LGTM from me.

File: modules/@angular/core/src/metadata/ng_module.ts [hunk 4]
```diff
@@ -177,6 +178,12 @@ export class NgModuleMetadata extends InjectableMetadata implements NgModuleMeta
    */
   schemas: Array<SchemaMetadata|any[]>;
 
+  /**
+   * An opaque ID for this module, e.g. a name or a path. Used to identify modules in `getNgModule`.
+   * If left `undefined`, the `NgModule` will not be registered with `getNgModule`.
+   */
+  id: string|undefined;
```
Comment (IgorMinar) [inline modules/@angular/core/src/metadata/ng_module.ts]: remove `|undefined` as that is not compatible with typescript 1.8

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts [hunk 5]
```diff
@@ -15,3 +15,32 @@ import {NgModuleFactory} from './ng_module_factory';
 export abstract class NgModuleFactoryLoader {
   abstract load(path: string): Promise<NgModuleFactory<any>>;
 }
+
+let deferreds: {[id: string]: NgModuleFactory<any>} = {};
```
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: why did you call this `deferreds`? there is nothing deferred about the content of this object. Maybe ngModuleFactoryRegistry?

Also can you use a `Map` instead of an object?

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts [hunk 6]
```diff
@@ -15,3 +15,32 @@ import {NgModuleFactory} from './ng_module_factory';
...
+  deferreds = {};
+}
+
+/**
+ * Returns the NgModuleFactory with the given id, if it exists and has been loaded.
+ * Modules that do not specify an `id` cannot be retrieved. Returns `null` if no such module exists (yet).
+ * @experimental
+ */
+export function getNgModule(id: string): NgModuleFactory<any> {
+  if (!deferreds[id]) throw new Error(`Module ${id} not loaded`);
```
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: ```new Error(`No NgModule with id '${id}' has not been loaded.`)```

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts [hunk 7]
```diff
@@ -15,3 +15,32 @@ import {NgModuleFactory} from './ng_module_factory';
...
+  deferreds[id] = factory;
+}
+
+export function clearModulesForTest() {
+  deferreds = {};
+}
+
+/**
+ * Returns the NgModuleFactory with the given id, if it exists and has been loaded.
+ * Modules that do not specify an `id` cannot be retrieved. Returns `null` if no such module exists (yet).
```
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: doesn't return `null` - you are throwing an error

File: modules/@angular/core/src/metadata/ng_module.ts (hunk 4 above)
Comment (mprobst) [inline modules/@angular/core/src/metadata/ng_module.ts]: Done.

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 6 above)
Comment (mprobst) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: Done (though w/o double negation :-))

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 3 above)
Comment (mprobst) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: Done.

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 7 above)
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: "Modules that do not ..." -> "Factories for NgModules that do not ..."

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 5 above)
Comment (mprobst) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: Ack, the API has grown a bit from a first attempt based on promises and deferreds. Done.

File: modules/@angular/compiler/src/identifiers.ts (hunk 2 above)
Comment (mprobst) [inline modules/@angular/compiler/src/identifiers.ts]: Done.

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts [hunk 8]
```diff
@@ -15,3 +15,32 @@ import {NgModuleFactory} from './ng_module_factory';
...
+ * @experimental
+ */
+export function getNgModule(id: string): NgModuleFactory<any> {
```
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: you are not getting a module but the factory of the module, so if anything this should be called `getNgModuleFactory`.

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 7 above)
Comment (mprobst) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: It's not the factory that declares the id, it's the NgModule (or its annotation).

Comment (mprobst) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: Done re returning null.

File: modules/@angular/compiler/src/identifiers.ts [hunk 9]
```diff
@@ -107,6 +107,11 @@ export class Identifiers {
...
+    runtime: registerLoadedModule,
```
Comment (IgorMinar) [inline modules/@angular/compiler/src/identifiers.ts]: @tbosch since we are passing in the runtime reference, this should work in the JIT mode with umd bundles were deep imports won't work. right?

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 8 above)
Comment (mprobst) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: Done.

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts [hunk 10]
```diff
@@ -15,3 +15,32 @@ import {NgModuleFactory} from './ng_module_factory';
 export abstract class NgModuleFactoryLoader {
   abstract load(path: string): Promise<NgModuleFactory<any>>;
 }
+
```
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: rather than implementing this as three different functions, why not implement it as static methods on the NgModuleFactoryLoader class? that way the functions are easier to document, discover and we don't pollute the api surface too much. just mark the individual methods as experimental.

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 6 above)
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: :) refactoring error

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 7 above)
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: "Factories for NgModules that do not specify an `id`" means that the NgModule doesn't specify the id. Why is that unclear? I also want to avoid us using term "module" to avoid confusing with ES6 modules.

File: modules/@angular/core/src/linker/ng_module_factory_loader.ts (hunk 10 above)
Comment (mprobst) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: Only one function is exported (`getModuleFactory`), the rest is private. User code will replace `NgModuleFactoryLoader` with their own implementation, so I think it's better to keep registration off of the class. It's odd to replace an implementation but then still refer to it, and `getModuleFactory` has no strong relationship with the loader instance.

Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: both are ugly, but since we have to rely on global state here we can't do much better. I missed that the clear method was not reexported. can you please update the public api spec so that we can see the final impact on the public api? (`gulp public-api:update`)

Comment (mprobst) [inline modules/@angular/core/src/linker/ng_module_factory_loader.ts]: Done.
